provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,218 @@
+		"name",
+		"id",
+		"arn",
+	}, stableOutputsNameOnly)
+
+	// we can't handle arrays because we don't know which item in the array
+	// the value corresponds to
+	if isArrayProperty(propName) {
+		return false
+	}
+
+	// e.g. aws-native:s3:StorageLens has a storageLensConfiguration/StorageLensArn readOnly property
+	if strings.Contains(propName, "/") {
+		parts := strings.Split(propName, "/")
+		name := parts[len(parts)-1]
+		// If the property is a nested property then only consider it stable if
+		// the property contains the resource type name. There are a lot of cases where
+		// an object has a property called `id` or `arn` that is not  the resource id or arn
+		return slices.Contains(stableOutputsNameOnly, naming.ToSdkName(name))
+	}
+	return slices.Contains(topLevelStableOutputs, naming.ToSdkName(propName))
+}
+
+func isReadOnly(readOnly []string, key string) bool {
+	for _, prop := range readOnly {
+		if naming.ToSdkName(prop) == key {
+			return true
+		}
+	}
+	return false
+}
+
+func filterReadOnly(readOnly []string, key string) []string {
+	result := []string{}
+	for _, prop := range readOnly {
+		if strings.HasPrefix(prop, key+"/") {
+
+			result = append(result, strings.TrimPrefix(prop, key+"/"))
+		}
+	}
+	return result
+}
+
+func isArrayProperty(prop string) bool {
+	return strings.Contains(prop, "/*/")
+}
+
+func arrayReadOnly(readOnly []string) []string {
+	var result []string
+	for _, prop := range readOnly {
+		result = append(result, strings.TrimPrefix(prop, "*/"))
+	}
+	return result
+}
+
+// previewOutputValue exists to recurse through nested objects and populate computed values
+// There are cases where the input and output types are the same, but some properties of the type
+// are only output values.
+func previewOutputValue(
+	inputValue resource.PropertyValue,
+	types map[string]metadata.CloudAPIType,
+	prop *schema.TypeSpec,
+	readOnly []string,
+) resource.PropertyValue {
+	switch {
+	case inputValue.IsArray() && (prop.Type == "array" || prop.Items != nil):
+		var items []resource.PropertyValue
+		for _, item := range inputValue.ArrayValue() {
+			items = append(items, previewOutputValue(item, types, prop.Items, arrayReadOnly(readOnly)))
+		}
+		return resource.NewArrayProperty(items)
+	case inputValue.IsObject() && strings.HasPrefix(prop.Ref, "#/types/"):
+		typeName := strings.TrimPrefix(prop.Ref, "#/types/")
+		if t, ok := types[typeName]; ok {
+			v := previewResourceOutputs(inputValue.ObjectValue(), types, t.Properties, readOnly)
+			return resource.NewObjectProperty(v)
+		}
+	case inputValue.IsObject() && prop.AdditionalProperties != nil:
+		inputObject := inputValue.ObjectValue()
+		result := resource.PropertyMap{}
+		for name, value := range inputObject {
+			p := value
+			result[name] = previewOutputValue(p, types, prop.AdditionalProperties, readOnly)
+		}
+		return resource.NewObjectProperty(result)
+	}
+	// fallback to assuming input is the same as the output
+	return inputValue
```

Comment:
This transformation is incomplete in handling the possible grammar cases of resource.PropertyValue, the ones I'm thinking about may be worth calling out:

- NewSecretProperty
- NewOutputProperty

More of these are in https://github.com/pulumi/pulumi/blob/5d73d64185ad32da8787a0e06875ecc59a912b11/sdk/go/common/resource/properties.go#L246

Also noting that the fallback is a little aggressive as it assumes the input value will be left unmodified. 

Perhaps there's something in the provider that assures us this code will not encounter NewSecretProperty or NewOutputProperty, but if it does, values wrapped in NewSecretProperty will not receive unknown marks for readOnly sub-properties but will be overly confident.

Reply:
I can add handling for secrets and outputs by just unwrapping them and calling `previewOutputValue` again with the unwrapped value.

As a side note, I noticed that in `aws-native` when we use `plugin.UnmarshalProperties` we pass `KeepUnknowns: true, KeepSecrets: true`. It doesn't seem like this is consistent across providers, do you know why?